pkg/smokescreen/conntrack/conn_tracker.go

```diff
@@ -15,19 +17,92 @@ type Tracker struct {
 	Wg           *sync.WaitGroup
 	statsc       statsd.ClientInterface
 
+	SuccessRateTracker *ConnSuccessRateTracker
+
 	// A connection is idle if it has been inactive (no bytes in/out) for this
 	// many seconds.
 	IdleTimeout time.Duration
 }
 
-func NewTracker(idle time.Duration, statsc statsd.ClientInterface, logger *logrus.Logger, sd atomic.Value) *Tracker {
+// ConnSuccessRateTracker tracks statistics about the overall success rate
+// of connection attempts over some time interval (which is set in StartNewConnSuccessRateTracker()).
+//
+// It tracks only the *most recently seen* connection to an individual destination host within the configured
+// time interval, to prevent a single destination host from having an outsized
+// impact on statistics.
+type ConnSuccessRateTracker struct {
+	ConnAttempts *cache.Cache
+	// The success rate of connection attempts, calculated over
+	// ConnAttempts over the interval configured in StartNewConnSuccessRateTracker()
+	ConnSuccessRateStats atomic.Value
+}
+
+// ConnSuccessRateStats represents a timestamped output of computations performed over
+// connection attempts.
+type ConnSuccessRateStats struct {
+	CalculatedAt    time.Time
+	ConnSuccessRate float64
+	TotalConns      int
+}
+
+// StartNewConnSuccessRateTracker creates a new ConnSuccessRateTracker with a specific calculation interval at which
+// ConnSuccessRateStats will be recomputed, and a time window to calculate those statistics over.
+func StartNewConnSuccessRateTracker(calculationInterval time.Duration, calculationWindow time.Duration) *ConnSuccessRateTracker {
+	newSuccessTracker := &ConnSuccessRateTracker{
+		ConnAttempts: cache.New(calculationWindow, time.Second*10),
```

Comment:
Instead of hardcoding the cleanup interval (the second arg) here, you might use a constant with a comment attached to it. How often cache entries get evicted seems relevant to the math you're doing here, or at least worth explicitly taking into consideration.

Reply:
To clarify, this parameter isn't the expiry deadline, it's just how quickly stuff gets removed from memory after it expires. It shouldn't have an impact on the calculation, moreso on memory usage. That being said, I think it makes sense to make it a parameter, so if we do want to tweak it we don't need a whole new smokescreen version!